### ISSUE TITLE
rfc for `deploymentId` style workerpool cycling

### ISSUE THREAD

**Issue by imbstack** [taskcluster/taskcluster#2891]
This is useful for being used by worker-runner to decide if it needs to terminate cloud workers or change config in static ones

Previous discussion in https://bugzilla.mozilla.org/show_bug.cgi?id=1593482

**Comment by imbstack** [comment]
The actual implementation of this should be easy so marking it as a small task. Figuring out exactly what we want to do first might take a few more days than that

**Comment by tomprince** [comment]
From [bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1593482#c7)
> Also, worth thinking about how this would interact with ci-admin/tc-admin. Would they update the deploymentId on every write? On no writes? Have some logic for it?

My first thought (for ci-admin) is to use the image name from [worker-images.yml](https://hg.mozilla.org/ci/ci-configuration/file/tip/worker-images.yml), but that doesn't handle the case of changing instance types.

**Comment by tomprince** [comment]
I was thinking about this a bit, and I wondered if the thing that makes sense to track is the launch-config associated to a given worker. And then worker-runner can terminate an instance if the launch-config that created the worker is no longer one of the launch-configs configured for the pool.

It feels a little bit like the idea of a deployment is trying to be an approximation of this, but in a way that requires other tools to mange the logic around it.

There are probably some settings that can be ignored in the launch-config, perhaps `utility` if is resurrected in #2879, and perhaps some provider specific values that affect provisioning but not the instance.

Keeping the launch-config around might also be useful for providing the metadata that [Bug 1641675](https://bugzilla.mozilla.org/show_bug.cgi?id=1641675) is looking for.

**Comment by petemoore** [comment]
There is quite a lot of context and history to this topic, which I'll try to present here:

* The concept of `deploymentId` was introduced [4 years ago](https://bugzilla.mozilla.org/show_bug.cgi?id=1298010#c9), as a means to inform workers that they are running in an outdated configuration
* Prior to this solution, several ideas were considered whereby workers would use existing data to decide whether they were outdated, and terminate/refresh their configuration accordingly
* Due to the variance of conditions that should lead to a worker refresh/termination across disparate worker pools and taskcluster deployments, it was decided that a general-purpose field to explicitly signal intent fore worker pool refresh was appropriate. This separates the worker termination/refresh decision making machinery from the worker and worker manager, and gives it to the worker manager config updating machinery, which in most cases will be ci-admin/tc-admin but also allows e.g. servo/deepspeech/other community projects to impose their own termination/refresh conditions which may be more appropriate for their projects/worker pools.

To give an example of potential differences in approach that may be warranted (and therefore the limitation of not using an explicit field), consider a worker pool is changed to use a cheaper instance type. Should existing workers be terminated?

* If it is purely for cost cutting measures, probably yes.
* However, perhaps this particular worker pool has a a high latency cost for starting new workers, further exacerbated by cold caches on new instances. Perhaps it is cheaper and more efficient, and does no harm to leave existing workers running.
* Perhaps the bigger instance type has a different gpu configuration, and tasks do not run properly on it, then we definitely want to flush the pool, and start with new instances.

Other examples:
* What if worker configuration changes? Perhaps some worker implementations can update their configuration on-the-fly without needing to terminate, and new workers spawn, and have cold caches.
* What about for optional configuration changes? Perhaps a config setting relates to a default PATH environment variable on the worker, and someone spotted that the PATH entries were not in consistent case (e.g. C:\Windows\....;c:\windows\....) and just wanted to clean that up for clarity. Perhaps there are thousands of running workers, and it isn't worth the financial cost ofrecycling the entire pool just for this superficial aesthetic change, but it is still worth making the change so that in a few days when all workers have rotated, they all have a cleaner PATH.

The point here is there are a multitude of possible reasons that someone may wish for a given worker pool to be recycled, or not recycled, for some given change, and there is no generic way to know whether that is the case or not. By providing the explicit choice to the "deployer" of whether that pool should be recycled or not, you create a more general purpose framework to cover a greater range of deployment scenarios that empowers users to apply the conditions they want. And for us, we can make blanket decisions about how to cycle our own worker pools by implementing that logic in ci-admin/tc-admin, rather than implement it directly in the platform.

For now I think we have a reasonable strategy of hashing the worker management configuration, and using that as an identifier for the deployment, assuming if any config changes at all, that the worker pool can be cycled. But other users can update the field in a different way if they have subtly different requirements.

**Comment by petemoore** [comment]
So a few things to bring up here:

If the purpose of `deploymentId` is purely to signal whether a worker pool should be gracefully recycled or not, perhaps we should instead just have a dedicated endpoint for that. Then when `tc-admin`/`ci-admin` makes worker pool updates, it could also call the endpoint to cycle the pool, based on some input (for example, based on an implicit instruction in the merge commit commit message).

It only makes sense to me for the concept of deployment identifier to exist directly in the taskcluster platform, if it has bearing on how the services operate, or if it affects business logic of the platform. If it is a transparent property for the services that only has value to the worker implementation, it doesn't make as much sense for it to be managed by the platform as a native platform feature.

If we choose to keep `deploymentId`, I believe it should be either be a peer of workerConfig, or inside it. I think if we only want one `deploymentId` per worker pool (instead of one per region etc) then we should also only have one workerConfig per worker pool too. As soon as worker configs have the ability to vary per region, it doesn't make sense that the `deploymentId` can't - although I'm not aware of region-specific worker configs, other than for setting `TASKCLUSTER_WORKER_LOCATION` - which is now implemented in workers directly, so doesn't warrant a dedicated region-specific config.

The nice thing about having a `deploymentId` property rather than needing to cycle worker pools is that you avoid the race condition of updating a worker pool definition, then new workers starting up with the new configuration, and then the API call being made to recycle the pool, when some of the workers are already running with the new configuration.

Regarding making `deploymentId` a feature that worker manager is aware of, I think dustin has highlighted the most obvious use case, which is being able to display the `deploymentId` of live workers in the UI.

As I understand it, the purpose of this issue is to cover two goals:
1) generic-worker already supports `deploymentId` natively, but `worker-runner` does not yet support it, so this helps us recover features that have been lost since migrating workers to worker-runner.
2) docker-worker does not have a means to cycle worker pools, so this brings the existing generic-worker feature to docker-worker worker pools.

Since we don't version entities in taskcluster (we do not keep a record of changes to worker pools etc) I'm not sure whether it adds a lot of value to display historic deploymentIds when they cannot be linked or associated to any other data, and they are mostly opaque identifiers.

If we insist they are a timestamp, that also limits the use of hashes of e.g. sha256 hashes of configs etc, as used by servo and others to generate the deploymentId.

To summarise, having considered all the options, at the moment it feels like the most logical thing to do is to simply port the existing feature as it is to worker-runner, and leave it in the worker configuration section of the worker pool config, not to be interpreted by the taskcluster services at all, but only to have relevance for software running on the worker (i.e. worker-runner in this case). Since the deploymentId feature already exists, and works, and we have users using it, perhaps the best is just to move it to worker runner as it is now in generic-worker, so that docker-worker can benefit from it too.

I'll try to canvass opinion from people today.

**Comment by SimonSapin** [comment]
I haven’t read the whole thread. But since I was asked, one data point is that for Servo’s CI we have:

* One pool of static (is that the term?) macOS workers on hardware that we manage. `deploymentId` is not used there.
* Some Linux / GCP managed in https://github.com/mozilla/community-tc-config/. I don’t know if or how `deploymentId` is used there
* One Windows / AWS pool that we manager with a script that wraps `tc-admin`. We set `deploymentId` to a hash of some of the configuration. https://github.com/servo/taskcluster-config/blob/818d3db243c886185c1744aade4d4323bfbb27b3/commands/tc.py#L114-L120 I think this system works fine and does what we need. Changing it fundamentally would create work to port this script to the new system

**Comment by petemoore** [comment]
After some further thought, and discussion with users, I'm leaning towards us not making any user-impacting changes here. The existing feature is a very simple and flexible mechanism, that provides the freedom for users to implement their own strategy for cycling worker pools, based on their own requirements. We should just make sure that it continues to work under worker-runner, so that it can be extended to docker-worker worker pools. Probably the best way to do this, in the least disruptive way, is to keep it as a generic-worker config setting, and add it as a docker-worker config setting with the same name. Worker runner already has custom logic for both generic-worker and docker-worker, so can take the config setting from the appropriate place.

This way, users aren't impacted, the feature becomes available to docker-worker, and we backfill one of the feature gaps introduced by the adoption of worker-runner on generic-worker.

**Comment by tomprince** [comment]
> If we choose to keep deploymentId, I believe it should be either be a peer of workerConfig, or inside it. I think if we only want one deploymentId per worker pool (instead of one per region etc) then we should also only have one workerConfig per worker pool too. As soon as worker configs have the ability to vary per region, it doesn't make sense that the deploymentId can't - although I'm not aware of region-specific worker configs, other than for setting TASKCLUSTER_WORKER_LOCATION - which is now implemented in workers directly, so doesn't warrant a dedicated region-specific config.

`workerConfig` is more likely to vary with instance type that region, but worker manager doesn't differentiate between varying things in thoses axes. For example, the `gecko-t/t-linux-metal` worker has several launch configs that differ in the `capacity` passed to `workerConfig`, based on instance size.

**Comment by imbstack** [comment]
The work for this issue is now going to be generating an rfc for this feature.

**Comment by petemoore** [comment]
Moved to https://github.com/taskcluster/taskcluster-rfcs/pull/161
